### PR TITLE
promote_consts: some clean-up after experimenting

### PR DESCRIPTION
**Pull request by tesuji**
This is some clean-up after experimenting in #125916,
Prefer to review commit-by-commit.

### CODE REVIEW

File: library/panic_unwind/Cargo.toml [hunk 1]
```diff
@@ -16,7 +16,7 @@ alloc = { path = "../alloc" }
 core = { path = "../core" }
 unwind = { path = "../unwind" }
 compiler_builtins = "0.1.0"
-cfg-if = "1.0"
+cfg-if = { version = "1.0", features = ['rustc-dep-of-std'] }
```
Comment (compiler-errors) [inline library/panic_unwind/Cargo.toml:19]: what is this change for?

Comment (tesuji) [reply library/panic_unwind/Cargo.toml:19]: When building with `x test test/mir-opt`, I noticed that std keeps rebuilding after no changes in it.
Setting `CARGO_LOG=cargo::core::compiler::fingerprint=info`, it reveals that cfg-if built different
for panic_unwind and std.
Having said that, I try again with simple `x build libary` multiple times without this change,
std is not getting rebuilt when unchanged.

Comment (tesuji) [reply library/panic_unwind/Cargo.toml:19]: I'm happy to remove this diff if it adds noice.

Comment (saethlin) [reply library/panic_unwind/Cargo.toml:19]: That's a good find, we should keep this change.

File: compiler/rustc_mir_transform/src/promote_consts.rs [hunk 2]
```diff
@@ -83,23 +84,21 @@ enum TempState {
...
-/// of a larger candidate.
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-struct Candidate {
-    location: Location,
-}
+/// Each `[Location]` is a "root candidate" for promotion, which will become
+/// the returned value in a promoted MIR, unless it's a subset of a larger
+/// candidate.
+#[derive(Clone, Debug)]
+struct Candidates(Vec<Location>);
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/promote_consts.rs:91]: I'm not convinced by this change. `Candidate` is a newtype to encode the invariant that the `Location` points to a "root candidate". Using a `Location` everywhere defeats that purpose.

Comment (tesuji) [reply compiler/rustc_mir_transform/src/promote_consts.rs]: It makes sense. Reverted!

File: compiler/rustc_mir_transform/src/promote_consts.rs [hunk 3]
```diff
@@ -58,9 +58,9 @@ impl<'tcx> MirPass<'tcx> for PromoteTemps<'tcx> {
         }
 
         let ccx = ConstCx::new(tcx, body);
-        let (mut temps, all_candidates) = collect_temps_and_candidates(&ccx);
+        let (mut temps, mut all_candidates) = collect_temps_and_candidates(&ccx);
 
-        let promotable_candidates = validate_candidates(&ccx, &mut temps, &all_candidates);
+        let promotable_candidates = validate_candidates(&ccx, &mut temps, &mut all_candidates);
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/promote_consts.rs:63]: If `all_candidates` is never used again, we should just move it to the callee.

**Pull request merged**